### PR TITLE
Laboratory work #4, Maksimova Ann - 19FPL2

### PR DESCRIPTION
**Pull request by amksmva**

### CODE REVIEW

File: lab_4/main.py [hunk 1]
```diff
@@ -2,46 +2,141 @@
...
 def tokenize_by_sentence(text: str) -> tuple:
-    pass
+    if not isinstance(text, str):
+        raise ValueError
+
+    text = re.sub(r'[^A-Za-z\n.?! ]', '', text).lower()
+    text = re.sub(r'[!?.]', ' <END> ', text)
+
+    tokens = text.split()
+    if tokens and tokens[-1] != '<END>':
```
Comment (WhiteJaeger) [inline lab_4/main.py:17]: Are you sure about `tokens[-1] != '<END>'`?

File: lab_4/main.py [hunk 2]
```diff
@@ -2,46 +2,141 @@
...
 
 def encode_text(storage: WordStorage, text: tuple) -> tuple:
-    pass
+    if not isinstance(storage, WordStorage) or not isinstance(text, tuple):
+        raise ValueError
+
+    encoded_text = []
+    for token in text:
+        id_token = storage.get_id(token)
+        encoded_text.append(id_token)
```
Comment (WhiteJaeger) [inline lab_4/main.py:69]: Can be concatenated into one line.

File: lab_4/main.py [hunk 3]
```diff
@@ -2,46 +2,141 @@
...
-        pass
+        if not isinstance(context, tuple):
+            raise ValueError
+
+        id_with_end = self._word_storage.get_id('<END>')
+
+        sentence = list(context)
+
+        attempts = 0
+        while attempts < 20:
```
Comment (WhiteJaeger) [inline lab_4/main.py]: If there is a finite number of times you are going to do an action -> another type of loops is better.

File: lab_4/main.py [hunk 4]
```diff
@@ -2,46 +2,141 @@
...
     def generate_text(self, context: tuple, number_of_sentences: int) -> tuple:
-        pass
+        if not isinstance(context, tuple) or not isinstance(number_of_sentences, int):
+            raise ValueError
+
+        text = ()
+        sentence = self._generate_sentence(context)
+        text += sentence
+        attempts = 1
+        while attempts < number_of_sentences:
```
Comment (WhiteJaeger) [inline lab_4/main.py]: If there is a finite number of times you are going to do an action -> another type of loops is better.

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]

File: lab_4/main.py (hunk 1 above)
Comment (amksmva) [reply lab_4/main.py:17]: It is written for the text == 'I am Ann'. (The sentence doesn't have the dot or smth else)

**Pull request closed without merging**